test/cpp/jit/test_gpu.cpp

```diff
@@ -19688,6 +19688,456 @@ TEST_F(NVFuserTest, FusionIssue1305Repro_CUDA) {
   TORCH_INTERNAL_ASSERT(t3->getComputeAtPosition() == 1);
 }
 
+TEST_F(NVFuserTest, FusionDoubleBuffering1_CUDA) {
+  Fusion fusion;
+  FusionGuard fg(&fusion);
+
+  auto tv0 = makeContigTensor(1);
+  fusion.addInput(tv0);
+
+  auto tv1 = set(tv0);
+  auto tv2 = add(tv1, IrBuilder::create<Double>(1.0));
+  auto tv3 = set(tv2);
+  fusion.addOutput(tv3);
+
+  tv1->doubleBuffer();
+  tv1->setMemoryType(MemoryType::Shared);
+
+  tv3->split(-1, 128);
+  tv3->split(-1, 32);
+  TransformPropagator::from(tv3);
+
+  tv0->computeAt(tv3, 1);
+
+  tv3->axis(-2)->parallelize(ParallelType::BIDx);
+  tv3->axis(-1)->parallelize(ParallelType::TIDx);
+  scheduler_utils::parallelizeAllLike(tv3, ir_utils::allTvs(&fusion));
+
+  FusionExecutor fe;
+  fe.compileFusion(&fusion);
+
+  auto options = at::TensorOptions().dtype(at::kFloat).device(at::kCUDA, 0);
+  at::manual_seed(0);
+  auto t0 = at::randn({1000}, options);
+  auto cg_outputs = fe.runFusion({t0});
+
+  auto ref = t0 + 1;
+
+  testValidate(&fusion, cg_outputs, {t0}, {ref}, __LINE__, __FILE__);
+}
+
+TEST_F(NVFuserTest, FusionDoubleBuffering2_CUDA) {
+  Fusion fusion;
+  FusionGuard fg(&fusion);
+
+  auto tv0 = makeContigTensor(1);
+  fusion.addInput(tv0);
+
+  auto tv1 = set(tv0);
+  auto tv2 = add(tv1, IrBuilder::create<Double>(1.0));
+  auto tv3 = set(tv2);
+  fusion.addOutput(tv3);
+
+  tv1->doubleBuffer();
+
+  tv3->split(-1, 128);
+  tv3->split(-1, 32);
+  TransformPropagator::from(tv3);
+
+  tv0->computeAt(tv3, -1);
+
+  tv3->axis(-2)->parallelize(ParallelType::BIDx);
+  tv3->axis(-1)->parallelize(ParallelType::TIDx);
+  scheduler_utils::parallelizeAllLike(tv3, ir_utils::allTvs(&fusion));
+
+  FusionExecutor fe;
+  fe.compileFusion(&fusion);
+
+  auto options = at::TensorOptions().dtype(at::kFloat).device(at::kCUDA, 0);
+  at::manual_seed(0);
+  auto t0 = at::randn({1000}, options);
+  auto cg_outputs = fe.runFusion({t0});
+
+  auto ref = t0 + 1;
+
+  testValidate(&fusion, cg_outputs, {t0}, {ref}, __LINE__, __FILE__);
+}
+
+TEST_F(NVFuserTest, FusionDoubleBuffering3_CUDA) {
+  Fusion fusion;
+  FusionGuard fg(&fusion);
+
+  auto tv0 = makeContigTensor(1);
+  fusion.addInput(tv0);
+
+  auto tv1 = add(tv0, IrBuilder::create<Double>(1.0));
+  auto tv2 = set(tv1);
+  auto tv3 = add(tv2, IrBuilder::create<Double>(1.0));
+  fusion.addOutput(tv3);
+
+  tv1->setMemoryType(MemoryType::Shared);
+
+  tv2->doubleBuffer();
+
+  tv3->split(-1, 128);
+  tv3->split(-1, 32);
+  TransformPropagator::from(tv3);
+
+  tv0->computeAt(tv3, 1);
+
+  // tv2 is invalid to double-buffer as its producer, tv1, is
+  // computed inside the double-buffering loop.
```

Comment:
In my understanding of double buffering, it is required. To be more precise, to double-buffer the load from `tv1` to `tv2`, `tv1` must be defined for the iteration domain where `tv2` is double buffered. That's why the CA position of `tv1` must be left of that of `tv2`. The double buffer axis of `tv2` can be further left if `axis(ca_pos - 1)` is parallelized or unrolled, and in that case the CA position of `tv1` must be further left.

What would the code look like with ca pos of tv1==ca pos of tv2?

Reply:
This seems to be from definition of double buffered iter domains, and I guess the ca pos could be moved by schedulers to fit in the requirement. 

Just wanted to understand more about scheduling under this requirement, mainly wanted to confirm if we could generate the below two cases:

My understanding for the `==` case is like: 
```cpp
  // T1[i, i(8)], T2[i, i(8)], both ca pos 1
for idx0 in ...
  float t2[8*2]; // double buffered
  float t1[8]
  for idx1 in 0..8:
    t1[idx1]=...
  for idx2 in 0..8:
    t2[idx2+8*(1-idx0%2)] = t1[idx2]

  ... (use t2[idx2+8*idx0%2])
```
This one could be seen very often and would be great if we could support.


ca pos of tv1 > ca pos of tv2 case could be 
```cpp
 // T1[i, i(8)] (ca -1), T2[i, i(8)] (ca 1)
for idx0 in ...
  float t2[8*2]; // double buffered
  for idx2 in 0..8:
   float t1[0];
   t1[0] = ... + ...
   t2[idx2+8*(1-idx0%2)] = t1[0]

  ... (use t2[idx2+8*idx0%2])
```
This might be useful in prolog fusion but seems low priority for now.

Reply:
In the first example, when assigning to `t2`, the logical indices of `t2` are `[idx0+1, idx2]`, and when `t2` is used, they are `[idx0, idx2]`, right? This means that the RHS of the assignment, `t1`, must also point to `[idx0+1, idx2]`, but `t1` is defined for `[idx0, idx2]`, so it doesn't work.

I did think about recursively changing the index of the direct and indirect producers of a double buffer tensor, but that seems like a challenge, and I'm not sure it's something we would really need.

In the first case, instead of double buffering `t2`, why not `t1`? If `t1` is defined with an input global memory tensor, double buffering `t1` should be fine.

Reply:
I see the challenge to do this in general here. Thanks for pointing this out. Would it be possible to support maybe just a limited set of cases? maybe just a chain of unary ops all the way back to gmem input would be all I need for now.

Reply:
Can you show an actual example code? If it's just a chain of unary ops leading to gmem, why not just double buffer the first load from the gmem?

Reply:
Tracking in #1397.